src/provider/view/utils/Grouping.ts

```diff
@@ -0,0 +1,101 @@
+import moment = require('moment');
+
+import {StoredNotification} from '../../model/StoredNotification';
+
+export enum GroupingType {
+    APPLICATION = 'Application',
+    DATE = 'Date'
+}
+
+/**
+ * A notification group - a set of notifications that should be grouped together
+ * within the UI.
+ *
+ * Grouping method can vary, but all grouping methods will produce a set of
+ * these groups.
+ */
+export interface Group {
+    /**
+     * Unique key for this group. Mainly used for storing a date string so as time passes we do not need to recalculate the date to group the notifications by.
+     */
+    key: string;
```

Comment:
Good catch. The key is used to group notifications.

When sorting by `Date`:
The `key` will be the `moment` date generated by `getGroupTitle`.

When sorting by `Application`
The `key` will be the `currentNotification.source.name || currentNotification.source.uuid;`

Added uuid to the name so it will apps with the same name don't merge.
```ts
let key: string = currentNotification.source.name + currentNotification.source.uuid || currentNotification.source.uuid;
```

Reply:
Would that snippet not group notifications by window, rather than application? I would've thought it'd just be `let key: string = currentNotification.source.uuid;`?..

But also, nothing is actually reading `key` - when deciding which group to place each notification in, we do the bucketing by user-facing title rather than the "behind the scenes" key - which makes `key` fairly redundant. If two apps had the same title, I think we'd end up with a situation where some of the notifications don't show in the center.

```
const groupTitle = getGroupTitle(currentNotification, groupMethod);
// If group title already exists just add it to the group
if (groups.has(groupTitle)) {
```

Reply:
This is fixed now as discussed.